### PR TITLE
Use SSE2 for single-channel audio resampling - 3x speedup.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It speeds up 1-channel audio resampling by a factor of 3 on x86 targets.

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * Linear samping of window function and its application using SSE2 intrinsics.
 - Affected modules and functionalities:
     * Audio resampling (decoder)
 - Key points relevant for the review:
     * N/A
 - Validation and testing:
     * Existing tests apply
 - Documentation (including examples):
     * N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/kernels/signal/resampling.h [hunk 1]
```diff
@@ -116,9 +135,25 @@ struct Resampler {
...
           i1 = n_in - 1 - in_block_i;
         float f = 0;
-        float x = i0 - in_pos;
-        for (int i = i0; i <= i1; i++, x++) {
-          assert(in_block_ptr + i >= in && in_block_ptr + i < in + n_in);
+        int i = i0;
+
+#ifdef __SSE2__
+        __m128 f4 = _mm_setzero_ps();
+        __m128 x4 = _mm_setr_ps(i - in_pos, i+1 - in_pos, i+2 - in_pos, i+3 - in_pos);
```
Comment (JanuszL) [inline dali/kernels/signal/resampling.h:142]: Can we assume that we have always at least 4 elements available?

File: dali/kernels/signal/resampling.h [hunk 2]
```diff
@@ -49,6 +52,22 @@ struct ResamplingWindow {
     return lookup[i] + di * (lookup[i + 1] - lookup[i]);
   }
 
+#ifdef __SSE2__
+  inline __m128 operator()(__m128 x) const {
+    __m128 fi = _mm_add_ps(x * _mm_set1_ps(scale), _mm_set1_ps(center));
```
Comment (JanuszL) [inline dali/kernels/signal/resampling.h:57]: Don't wee need to assure that `idx[3]+1 < lookup.size()`?

Comment (mzient) [reply dali/kernels/signal/resampling.h:57]: It's externally assured, just as it was for scalar variant that `i+1 < lookup.size()`. Or do you mean to add assert?

File: dali/kernels/signal/resampling.h (hunk 1 above)
Comment (mzient) [reply dali/kernels/signal/resampling.h:142]: If we don't, the vectorized loop won't get executed at all, so we're safe.

**Comment by mzient** [comment]
!build

File: dali/kernels/signal/resampling.h (hunk 2 above)
Comment (JanuszL) [reply dali/kernels/signal/resampling.h:57]: I'm rather asking if we need such assert. If not then I'm fine too.

Comment (mzient) [reply dali/kernels/signal/resampling.h:57]: The assert was there in the scalar variant for debugging. I don't remember if it ever fired, but I kept it there. I'll probably put one here too, if we decide to extend SIMD support to multi-channel audio (now only 1ch variant got the upgrade). I think that for most scenarios it's OK, but there certainly are some where multi-channel audio is required.

**Review by JanuszL — APPROVED** [review]

**Review by klecki — APPROVED** [review]

**Pull request merged**